Applications/SimpleShot/layers/centering.cpp

```diff
@@ -0,0 +1,538 @@
+  1.2456608,  1.5908684,  1.3463796,  1.781804,   1.2539327,  1.0488071,
+  0.69456387, 0.7837838,  1.420064,   0.96974546, 0.7254165,  0.91133595,
+  0.7480144,  1.6793756,  1.2010552,  0.37359616, 0.8411831,  0.94810975,
+  1.081522,   1.4002662,  0.91972107, 1.0329305,  1.4743291,  0.9981039,
+  1.0227196,  0.352504,   1.8154447,  1.3830396,  0.8657597,  1.4396735,
+  1.5980465,  0.99679625, 1.0142295,  1.4402367,  0.38848996, 0.9896359,
+  0.59641224, 1.1450837,  0.9463809,  0.618432,   0.6541309,  0.8063583,
+  1.6206524,  1.840674,   1.6155039,  1.1257526,  0.61734056, 0.6324227,
+  1.2998347,  2.1847007,  0.4944939,  1.3934348};
+
+namespace {
+nntrainer::Tensor getHardCodedFeatureVector(const std::string &model,
+                                            const std::string &dataset) {
+  if (model == "resnet50" && dataset == "cifar100") {
+    return nntrainer::Tensor({2048}, cifar100_resnet50);
+  }
+
+  if (model == "conv4" && dataset == "cifar100") {
+    return nntrainer::Tensor({64}, cifar100_conv4);
+  }
+
+  if (model == "ones" && dataset == "ones") {
+    nntrainer::Tensor t({4});
+    t.setValue(1);
+    return t;
+  }
+  std::stringstream ss;
+  ss << "current combination is not supported"
+     << " model: " << model << "dataset: " << dataset;
+  throw nntrainer::exception::not_supported(ss.str().c_str());
+}
+} // namespace
+
+const std::string CenteringLayer::type = "centering";
+
+CenteringLayer::CenteringLayer(const std::string &model,
+                               const std::string &dataset) :
+  Layer() {
+  mean_feature_vector = getHardCodedFeatureVector(model, dataset);
+}
+
+int CenteringLayer::setProperty(std::vector<std::string> values) {
+  util::Entry e;
+
+  std::vector<std::string> unhandled_values;
+
+  std::string model, dataset;
+
+  for (auto &val : values) {
+    try {
+      e = util::getKeyValue(val);
+    } catch (std::invalid_argument &e) {
+      std::cerr << e.what() << std::endl;
+      return -1;
+    }
+
+    if (e.key == "model") {
+      model = e.value;
+    } else if (e.key == "dataset") {
+      dataset = e.value;
+    } else {
+      unhandled_values.push_back(val);
+    }
+  }
+
+  if (model != "" && dataset != "") {
+    try {
+      mean_feature_vector = getHardCodedFeatureVector(model, dataset);
+    } catch (std::invalid_argument &e) {
+      std::cerr << e.what() << std::endl;
+      return ML_ERROR_INVALID_PARAMETER;
+    }
+  }
+
+  return nntrainer::Layer::setProperty(unhandled_values);
+}
+
+int CenteringLayer::initialize(nntrainer::Manager &manager) {
+  output_dim[0] = input_dim[0];
+
+  return ML_ERROR_NONE;
+}
+
+void CenteringLayer::forwarding(bool training) {
+  net_input[0]->getVariableRef().add(mean_feature_vector,
+                                     net_hidden[0]->getVariableRef(), -1);
+}
+
+void CenteringLayer::calcDerivative() {
+  // do nothing but pass back as this is inplacable operation
+  nntrainer::Tensor &derivative_ = net_hidden[0]->getGradientRef();
```

Comment:
I don't think we need this unless we are going to update the previous layer's weight with backwarding.

Reply:
This is a pure virtual function, so this layer needs a stub method at least.

Reply:
This is changed to noop